### PR TITLE
Draft: Add ppc64le pipelines for ML builds

### PR DESCRIPTION
**Pull request by nicholas-sly**

### CODE REVIEW

**Comment by adamjstewart** [comment]
I'd rather name everything "ppc64le" to be consistent with `spack arch`

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cuda/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cuda/spack.yaml]: Each stack needs its own mirror, and this url is already used for the existing ml cuda stack.  Also it makes one think the arch is x86_64 when it's actually ppc. 

edit: I just noticed you did this for the other mirror url, so just do the same thing here.  Maybe:

```
  mirror: s3://spack-binaries/develop/ml-linux-power64le-cuda
```

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: No need for this commented section I think, jobs associated with your stacks should be picking up existing linux/power configs thanks to how your new jobs extend the `.linux-power` and `.base-generate` definitions.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: If these aren't relevant to your stack, let's get rid of them.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: I think one of these `build-job` sections should be removed.  @kwryankrattiger can correct me when I get this backwards, but I think these are applied bottom up, so your final job configuration will just have the first one.  In which case, you can remove the second.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cuda/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cuda/spack.yaml]: Same thing as above wrt the repeated `build-job` sections.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: This image does not seem to exist. At least that's what it says [here](https://gitlab.spack.io/spack/spack/-/jobs/7843299#L10), and I can't seem to find it on dockerhub.

It seems to be mentioned elsewhere in the PR too, so I'll just point it out here.

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
Thanks for contributing some new stacks @nicholas-sly!  I've added a few comments, addressing the one about the missing docker image may let your stacks get into the actual pipeline generation step on gitlab.

Also, @eugeneswalker manages the only power runners available to us, let's see what he has to say about their capacity to handle this extra workload.

I pinged @kwryankrattiger in my review as well, since he came up with the new scheme for organizing the gitlab ci configs.

Thanks again!

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Yes, I have not been able to find an official list of what containers are available.  Todd pointed me at https://github.com/spack/gitlab-runners , but that doesn't specify the final image names and tags available, so I'm having to guess and check.  That page does seem to imply that there is a fedora36 image that supports ppc64le, but unclear what the exact name and tag are.  No other pipelines currently appear to use it either.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: So, I assume you mean only have one of the ubuntu or fedora jobs.  I'm actually intentionally targeting both as many of these packages, if they even support linux, are built for Ubuntu.  If users are installing these packages locally on a linux system, it is very likely Ubuntu.  But most of the HPC systems that I've interacted with are either SLES or RHEL-based.  As such, I think it is useful to ensure that these packages are tested against both OSes as best we can.

If I'm just specifying it incorrectly to achieve that end, then I'm happy to change it.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: These were used for local efforts at getting the stacks to concretize.  They will be removed before the PR is ready for final review, but as it is currently marked as `Draft`, I'll leave them until I'm sure they're not required.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: Also used for local testing.  Will be removed before PR is finalized.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (eugeneswalker) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: We maintain the `ecpe4s` repo on DockerHub and we've never made a fedora36 image by the name.

What we have at `ecpe4s` are the following images:

AMD64
* `ecpe4s/ubuntu20.04-runner-amd64-gcc-11.4:2023.08.01`
* `ecpe4s/ubuntu20.04-runner-amd64-gcc-12.3:2023.08.01`

PPC64LE
* `ecpe4s/ubuntu20.04-runner-ppc64le-gcc-11.4:2023.08.01`
* `ecpe4s/ubuntu20.04-runner-ppc64le-gcc-12.3:2023.08.01`

ARM64
* `ecpe4s/ubuntu20.04-runner-arm64-gcc-11.4:2023.08.01`
* `ecpe4s/ubuntu20.04-runner-arm64-gcc-12.3:2023.08.01`

Please see this repo for the recipes:
* https://github.com/UO-OACISS/e4s/tree/master/docker-recipes/runner

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: What about `ecpe4s/rhel8-runner-ppc64le:2023-01-01`?  That would seem to satisfy my need for a RHEL-based runner on power architecture.  It hasn't been updated since the new year, but I've worked with worse.  Also unclear if the recent RHEL licensing changes will affect the long-term usability of such an image.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Still some remaining places that refer to power64le instead of ppc64le

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: This is a generic error that means "something went wrong". The real error is farther up in the output. If you share a log with me I can take a look. We build PyTorch on ppc64le for another stack so it should definitely work.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Should this be uncommented now too?

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Can remove this comment

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: I was waiting to see how the build did with py-torch before uncommenting it.  It looks like the RHEL8 CPU build still failed.  Should probably take a look at that first.

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: nvm, the CI interface got weird after all the tests had run.  I guess that was a different pipeline.  I'll uncomment and remove as needed and push again.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Are these variants required?

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Personally, horovod doesn't seem to serve much purpose if it isn't built with MPI.  For the CPU build, NCCL doesn't make sense (does it?), so I stick with MPI.  Everything I've heard is that GLOO isn't all that well supported.  I think a framework is required, so I gave pytorch as it seems to be the most useful/compatible.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Looks like, for some reason, py-horovod still depends on py-tensorboard, even if it's using the pytorch framework and not tensorflow ( https://gitlab.spack.io/spack/spack/-/pipelines/480840 ).

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: pytorch is already the default framework and mpi is already the default controller. And `tensor_ops=nccl` should be disabled due to conflicts with `~cuda~rocm`. So I don't think this is needed.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-power64le-cpu/spack.yaml]: I've already experienced builds that succeeded with ubuntu but not with rhel8.  I'm leaving it in because tests that pass on ubuntu do us no good when the majority of HPC systems are on a rhel-based OS.  It really comes down to what our goal is with the CI system.  If we want to ensure that changes to these packages won't have a significant impact to most of our users that try and build these recipes on most of the systems they might use, then we need to ensure that our CI is representative of those systems and users.  If we just want a build that works even if it doesn't represent anything an end user might use, we can probably just ignore ppc64le altogether.  As for CI bandwidth, most changes in Spack shouldn't even trigger this.  The point (as I see it) is to test for any changes that will impact these builds.  Once these pipelines are integrated, they shouldn't be triggered often, unless they need to be triggered.  Not sure how we judge this bandwidth or if such limiting factors are publicly available, but I think this is necessary for now.  Obviously, once all of these systems have been decommissioned, we can remove these pipelines.  Until then, these pipelines are a useful tool for trying to get these ML packages working on the machines we want them to build on.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Seemingly irrelevant in the face of the fact that py-horovod fails to build even with these default variants as it seems to still depend on py-tensorboard-data-server, which has always failed to build on ppc64le.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: For some reason it's building py-pytorch-lightning 1.5.3 instead of the latest version. If you can figure that out it will no longer depend on tensorboard. But would be great to get tensorboard working.

**Comment by nicholas-sly** [comment]
@scottwittenburg @adamjstewart It seems the tests are all passing now.  Let me know if you have any further changes to request.  Github says there's one unaddressed request, but everything I'm seeing is either outdated or addressed.  Thanks.

**Comment by adamjstewart** [comment]
What's the reason for no ROCm?

**Comment by nicholas-sly** [comment]
> What's the reason for no ROCm?

Are there any machines that have ppc64le processors and AMD GPUs?  I'm not aware of any and I'm under the impression that IBM isn't going to be making new machines.  If that's not the case, we can try.

**Comment by adamjstewart** [comment]
I have no idea

**Review by adamjstewart — DISMISSED** [review]
We can figure out the Bazel stuff another day

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
Every pipeline needs to write into it's own mirror.  Concurrent pipelines writing to the same destination creates race conditions which result in weird errors.

Currently the solution is to have a `spack.yaml` for every pair of `generate`/`build` jobs.  But once we have #39939, that won't be necessary, as the `spack.yaml` will no longer be where we specify the mirrors for spack pipelines.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Did you mean to have some submapping here, rather than just `build-job`?  If this was working, I'm not sure how.  Given you were trying to have two `spack.yaml` for the four pipelines, you may have been able to do something like this instead:

Example:

```yaml
...
 - match_behavior: first
    submapping:
    - match:
      - os=ubuntu20.04
        build-job:
          image:
            name: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
            entrypoint: ['']
```

But that's irrelevant if you change to one stack per pipeline.  If you wait until #39939 is merged, then I think you could stick with only two `spack.yaml` for your four pipelines, and the above example would be how you map build jobs to docker image.

cc @kwryankrattiger

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Not sure what the issue is here.  I want each stack to build on both ubuntu and rhel8.  In yaml, the `pipeline-gen` key takes a list, so I just put two elements in that list.  You can see from the CI build that this properly broke out the pipeline into separate jobs, each utilizing the appropriate container.

I am planning to wait for #39939 to be merged to avoid breaking this into multiple files only to be able to deduplicate after that PR goes through.

Comment (scottwittenburg) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: It looks to me like it didn't break them out properly.  This [job](https://gitlab.spack.io/spack/spack/-/jobs/8227536) was supposed to be a rhel8, but ran on the ubuntu container.

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Interesting indeed.  The container does seem to indicate ubuntu for the rhel job.  But then the `uname -a` indicates a rhel OS.  Presumably that's the host, so fair enough.  The `spack arch` output seems to corroborate the ubuntu OS.  But then the `gnuconfig` dependency installation along with the libiconv installation directories indicate a rhel OS.  I guess that's due to the generate builds occurring with the correct container.  I just pushed a commit that tries to be more explicit with which container image should be used for the CI jobs.

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: For reasons I cannot fathom, a CI jobs ending in `-build` will not accept an `image` key in the `.gitlab-ci.yaml` file.  Likewise, it can "extend" another job that does take an `image` key, but does not honor that same `image` when building.  I can try modifying the `spack.yaml` files according to your above example, but without proper documentation, I'll be doing a good bit of guess and check with CI parsing to try and get it right.

Comment (scottwittenburg) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: > For reasons I cannot fathom, a CI jobs ending in -build will not accept an image key in the .gitlab-ci.yaml file.

Jobs ending in `-build`?  I guess the trigger jobs do seem to all end in `-build`, here's a [link](https://docs.gitlab.com/ee/ci/yaml/#trigger) to the gitlab docs describing what keys are valid in trigger jobs.

> but without proper documentation, I'll be doing a good bit of guess and check with CI parsing to try and get it right

This [bit](https://spack.readthedocs.io/en/latest/pipelines.html#ci-yaml) of the pipeline docs seems to illustrate an example of what you're trying to do.

Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: Forgot about this comment, but want to resurface this:

> I think testing both ubuntu and rhel is overkill right now. We don't even do that for x86_64. This is something worth adding someday but idk if we have the CI bandwidth to do both yet.

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: I stand by my response to that comment: https://github.com/spack/spack/pull/39174#discussion_r1309191617

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: The latest pipeline run https://gitlab.spack.io/spack/spack/-/pipelines/500642 is an example of exactly this happening.

Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: FYI, Spack is slowly moving in the direction of modeling glibc such that packages will no longer require _any_ system dependencies. Once this is done, Ubuntu and RHEL will be the same system, and there should be no real reason to test both.

Comment (nicholas-sly) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-ppc64le-cpu/spack.yaml]: While I appreciate your optimism in this respect, the demonstrable differences between the two OSes means I'm going to have to see the two pipelines succeed side-by-side with a significant portion of the ML packages we're interested in before I'm willing to down sample to a single OS.

**Comment by adamjstewart** [comment]
Based on Slack discussion, it seems like ppc64le is on its way out and our CI runners are limited anyway. Let's scrap this unless IBM decides to contribute to this effort.

**Pull request closed without merging**
